### PR TITLE
Docs: Explaining plugin machinery for contributors

### PR DESCRIPTION
**Pull request by dav009**
Signed-off-by: David Przybilla <dav.alejandro@gmail.com>

# TL;DR
Explaining Flyte plugin machinery for upcoming contributors

## Complete description
Documenting little bits that I've learnt through the office sessions and tiny contributions. 
I did not know what was the best place to link. Also Suggestions for a better title are welcome!

---

- [x] ran `make docs` successfully

### CODE REVIEW

**Comment by dav009** [comment]
@samhita-alla   🙏

**Comment by kumare3** [comment]
@dav009 this is amazing- absolutely Fantastic. Thank you for writing this up. Cc @EngHabu

File: rsts/community/contribute.rst [hunk 1]
```diff
@@ -139,6 +139,8 @@ Output:
 🧱 Component Reference
 ======================
 
+A good way to understand how components below interact with each other would be to take a look at :ref:`Understand the lifecycle of a workflow <lifecycle-workflow>`, which explains how these components interact with each other.
+
```
Comment (SmritiSatyanV) [inline rsts/community/contribute.rst:143]: To understand how components below interact with each other, refer to :ref:`Understand the lifecycle of a workflow <lifecycle-workflow>`.

File: rsts/concepts/lifecycle_workflow_with_plugin.rst [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+#################################################################
+Understand the Lifecycle of a Flyte Workflow
+#################################################################
+
+After reading this entry you should be able to understand how Flyte's Plugin machinery
+work under the hood and how information flows from the different
+components in Flyte.
+
+Under the hood Flyte relies on a primitive called “Plugins”. Every task
```
Comment (SmritiSatyanV) [inline rsts/concepts/lifecycle_workflow_with_plugin.rst:11]: Under the hood,

File: rsts/concepts/lifecycle_workflow_with_plugin.rst [hunk 3]
```diff
@@ -0,0 +1,252 @@
+.. _lifecycle-workflow:
+
+#################################################################
+Understand the Lifecycle of a Flyte Workflow
+#################################################################
+
+After reading this entry you should be able to understand how Flyte's Plugin machinery
+work under the hood and how information flows from the different
+components in Flyte.
+
```
Comment (SmritiSatyanV) [inline rsts/concepts/lifecycle_workflow_with_plugin.rst:10]: Let's understand how Flyte's Plugin machinery works and how information flows from one component to another in Flyte.

File: rsts/concepts/lifecycle_workflow_with_plugin.rst [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+Under the hood Flyte relies on a primitive called “Plugins”. Every task
+that you run on Flyte is somehow powered by a plugin. Some of those plugins are native and guaranteed by Flyte system. These native plugins are for example powering your python functions (i.e: Tasks) by running them inside a k8s pod. There are Three native plugins : ``Container``, ``K8sPod`` and ``Sql``.
+
```
Comment (SmritiSatyanV) [inline rsts/concepts/lifecycle_workflow_with_plugin.rst:13]: Under the hood, Flyte relies on a primitive called “Plugins”. Every task that you run on Flyte is powered by a plugin. Some of these plugins are native and guaranteed by Flyte system. These native plugins, for example, run your Python functions (such as tasks) inside a k8s pod. There are three native plugins: ``Container``, ``K8sPod`` and ``Sql``.

File: rsts/concepts/lifecycle_workflow_with_plugin.rst [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+Addionally there are also plugins which are actual extensions, they create additional infrastructure, talk to other SaaS on your behalf. Examples of those are the Spark or AWS Athena plugins.
+
+A plugin requires code to live in multiple locations
+
```
Comment (SmritiSatyanV) [inline rsts/concepts/lifecycle_workflow_with_plugin.rst:17]: There are plugins which are actual extensions; they create additional infrastructure, and communicate with SaaS on your behalf. Example: Spark, AWS Athena.

 A plugin requires code to live in multiple locations

Comment (SmritiSatyanV) [reply rsts/concepts/lifecycle_workflow_with_plugin.rst:17]: Can we add links to Spark and AWS Athena plugin?

File: rsts/concepts/lifecycle_workflow_with_plugin.rst [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+1. Some plugin logic has to live in SDK so that users can create tasks
+   that make use of a plugin. This logic usually goes into a specific
+   SDK i.e: flytekit’s python: https://github.com/flyteorg/flyteplugins
+
```
Comment (SmritiSatyanV) [inline rsts/concepts/lifecycle_workflow_with_plugin.rst:21]: Plugin logic lives in the SDK for users to create tasks that uses the plugin. This logic resides in the goes specific SDK, that is: flytekit’s python: https://github.com/flyteorg/flyteplugins.

File: rsts/concepts/lifecycle_workflow_with_plugin.rst [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+2. Another big chunk of plugins logic lives in
+   `Flyteplugins <https://github.com/flyteorg/flyteplugins>`__. This
+   repository is a library that gets loaded into
+   `FlytePropeller <https://github.com/flyteorg/flytepropeller>`__.
+   Flytepropeller is a Kubernetes operator, when Flytepropeller starts
```
Comment (SmritiSatyanV) [inline rsts/concepts/lifecycle_workflow_with_plugin.rst:26]: This is a library that is loaded into `FlytePropeller <https://github.com/flyteorg/flytepropeller>`__.

File: rsts/concepts/lifecycle_workflow_with_plugin.rst [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+   it loads Flyteplugins. Flytepropeller knows that plugins exists and that
+   executing tasks depends on them. However Flytepropeller knows nothing
+   about “how to execute plugins”.
+
```
Comment (SmritiSatyanV) [inline rsts/concepts/lifecycle_workflow_with_plugin.rst:30]: FlytePropeller (a Kubernetes operator) loads Flyteplugin upon starting.

File: rsts/concepts/lifecycle_workflow_with_plugin.rst [hunk 9]
```diff
@@ -0,0 +1,252 @@
+.. _lifecycle-workflow:
```
Comment (samhita-alla) [inline rsts/concepts/lifecycle_workflow_with_plugin.rst:1]:
```suggestion
.. _workflow-lifecycle:
```

Comment (samhita-alla) [reply rsts/concepts/lifecycle_workflow_with_plugin.rst:1]: Could you rename the file to workflow_lifecycle.rst?

File: rsts/concepts/lifecycle_workflow_with_plugin.rst [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+After reading this entry you should be able to understand how Flyte's Plugin machinery
+work under the hood and how information flows from the different
+components in Flyte.
+
+Under the hood Flyte relies on a primitive called “Plugins”. Every task
+that you run on Flyte is somehow powered by a plugin. Some of those plugins are native and guaranteed by Flyte system. These native plugins are for example powering your python functions (i.e: Tasks) by running them inside a k8s pod. There are Three native plugins : ``Container``, ``K8sPod`` and ``Sql``.
+
+Addionally there are also plugins which are actual extensions, they create additional infrastructure, talk to other SaaS on your behalf. Examples of those are the Spark or AWS Athena plugins.
+
+A plugin requires code to live in multiple locations
```
Comment (samhita-alla) [inline rsts/concepts/lifecycle_workflow_with_plugin.rst:16]: @dav009, could you explain your intention here? I don't understand this statement.

File: rsts/concepts/lifecycle_workflow_with_plugin.rst (hunk 6 above)
Comment (samhita-alla) [reply rsts/concepts/lifecycle_workflow_with_plugin.rst:21]: Are you referring to https://github.com/flyteorg/flytekit/tree/master/plugins? I don't think the link that has been attached is correct.

File: rsts/concepts/lifecycle_workflow_with_plugin.rst [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+   SDK i.e: flytekit’s python: https://github.com/flyteorg/flyteplugins
+
+2. Another big chunk of plugins logic lives in
+   `Flyteplugins <https://github.com/flyteorg/flyteplugins>`__. This
+   repository is a library that gets loaded into
+   `FlytePropeller <https://github.com/flyteorg/flytepropeller>`__.
+   Flytepropeller is a Kubernetes operator, when Flytepropeller starts
+   it loads Flyteplugins. Flytepropeller knows that plugins exists and that
+   executing tasks depends on them. However Flytepropeller knows nothing
+   about “how to execute plugins”.
```
Comment (samhita-alla) [inline rsts/concepts/lifecycle_workflow_with_plugin.rst:29]:
```suggestion
   FlytePropeller (a Kubernetes operator) loads Flyteplugin upon starting. 
   Flytepropeller knows that plugins exist and that
   executing tasks depends on them. However, Flytepropeller knows nothing
   about “how to execute the plugins”.
```

File: rsts/concepts/lifecycle_workflow_with_plugin.rst (hunk 6 above)
Comment (dav009) [reply rsts/concepts/lifecycle_workflow_with_plugin.rst]: @samhita-alla you are right, we should link  ` https://github.com/flyteorg/flytekit/tree/master/plugins`.

File: rsts/concepts/lifecycle_workflow_with_plugin.rst (hunk 10 above)
Comment (dav009) [reply rsts/concepts/lifecycle_workflow_with_plugin.rst]: @samhita-alla  I want to convey that a plugin code most likely implies some code living in the operator and some code living in the SDK, both of which are different repositories and/or programming languages.

**Comment by dav009** [comment]
@samhita-alla  @SmritiSatyanV thanks for your comments. I think I addressed most of them. Feel free to re-review

File: rsts/concepts/workflow_lifecycle.rst [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+Let's understand how Flyte's plugin machinery works and how information flows from one component to another in Flyte.
+
+Under the hood, Flyte relies on a primitive called “Plugins”. Every task that you run on Flyte is powered by a plugin. Some of these plugins are native and guaranteed by Flyte system. These native plugins, for example, run your Flyte tasks inside a k8s pod. There are three native plugins, namely, ``Container``, ``K8sPod``, and ``Sql``.
+
+Moreover, there are plugins that are actual extensions; they create additional infrastructure and communicate with SaaS on your behalf. Examples include :ref:`Spark <intermediate_using_spark_tasks>`, :ref:`AWS Athena <AWS Athena>`, etc.
+
+A plugin requires code to live in multiple locations.
+
+1. Some parts of plugin logic reside in the SDK for users to create tasks that use the plugin. This logic is present in a specific SDK, like Flytekit’s Python: https://github.com/flyteorg/flytekit/tree/master/plugins.
```
Comment (kumare3) [inline rsts/concepts/workflow_lifecycle.rst:15]: I would draw a corollary here. Say it's like a client for an RPG service

**Review by kumare3 — APPROVED** [review]

**Comment by kumare3** [comment]
RPC service

**Comment by dav009** [comment]
@kumare3  I was wondering what RPG was :'D

**Review by samhita-alla — APPROVED** [review]

**Pull request merged**